### PR TITLE
TokenFeeScheduleUpdate `handle()` implementation

### LINKED ISSUE #6383: Implement `handle` for TokenFeeScheduleUpdateHandler

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Fixes #6382 
Fixes https://github.com/hashgraph/hedera-services/issues/6547
Fixes https://github.com/hashgraph/hedera-services/issues/6358
- Implemented `TokenFeeScheduleUpdateHandler` , `handle()` method
- Added `CustomFeesValidator`  to validate custom fees . It will be shared between `TokenCreate` and `TokenFeeScheduleUpdate`
- Separated `ReadableTokenRelationStoreImpl`, so it can be created using `storeFactory`

### CODE REVIEW

File: hedera-node/hedera-token-service-impl/build.gradle.kts [hunk 1]
```diff
@@ -29,6 +29,7 @@ configurations.all {
 
 dependencies {
   implementation(project(":hedera-node:hapi"))
+  implementation(project(mapOf("path" to ":hedera-node:hedera-config")))
```
Comment (hendrikebbers) [inline hedera-node/hedera-token-service-impl/build.gradle.kts:32]:
```suggestion
  implementation(project(":hedera-node:hedera-config"))
```

Comment (hendrikebbers) [reply hedera-node/hedera-token-service-impl/build.gradle.kts:32]: @Neeharika-Sompalli why do you use that notation or is it generated somehow?

Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/build.gradle.kts:32]: @hendrikebbers When I add the needed import which needs dependency,  IntelliJ shows a suggestion. When I click it, it generates
```
implementation(project(mapOf("path" to ":hedera-node:hedera-config")))
```

**Review by hendrikebbers — DISMISSED** [review]

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java [hunk 2]
```diff
@@ -63,14 +81,67 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
...
+            @NonNull final HandleContext context,
+            @NonNull final TransactionBody txn,
+            @NonNull final WritableTokenStore tokenStore) {
+        requireNonNull(context);
+        requireNonNull(txn);
+        requireNonNull(tokenStore);
+
+        // get the latest configuration
+        final var config = context.getConfiguration().getConfigData(TokenServiceConfig.class);
+        var op = txn.tokenFeeScheduleUpdateOrThrow();
```
Comment (tinker-michaelj) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java:99]: Can be `final` 😛.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/ReadableTokenRelationStoreImpl.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    public ReadableTokenRelationStoreImpl(@NonNull final ReadableStates states) {
+        this.readableTokenRelState = requireNonNull(states).get(TokenServiceImpl.TOKEN_RELS_KEY);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Optional<TokenRelation> get(final long tokenNum, final long accountNum) {
```
Comment (mhess-swl) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/ReadableTokenRelationStoreImpl.java:52]: Again, switching the order of these in my PR to `accountNum, tokenNum`

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java (hunk 2 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java:99]: Done 🙂

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImplTest.java [hunk 4]
```diff
@@ -167,26 +166,6 @@ void testSuccess() throws PreCheckException, IOException {
     @Nested
     @DisplayName("0. Node state pre-checks")
     class NodeTests {
-        @Test
-        @DisplayName("When the node is zero stake, the transaction should be rejected")
-        void testParseAndCheckWithZeroStakeFails() throws IOException, PreCheckException {
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImplTest.java:172]: yes, Richard pointed it out

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java [hunk 5]
```diff
@@ -141,4 +141,9 @@ protected void finishUtilPrng(@NonNull final PrngRecordBuilder recordBuilder) {
             sideEffectsTracker.trackRandomBytes(PbjConverter.asBytes(recordBuilder.getPrngBytes()));
         }
     }
+
+    @Override
+    protected void finishTokenFeeScheduleUpdate(@NonNull final WritableTokenStore tokenStore) {
+        tokenStore.commit();
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java:147]: Done

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java [hunk 6]
```diff
@@ -38,16 +50,22 @@
  */
 @Singleton
 public class TokenFeeScheduleUpdateHandler implements TransactionHandler {
+    private final CustomFeesValidator customFeesValidator;
+
     @Inject
-    public TokenFeeScheduleUpdateHandler() {
-        // Exists for injection
+    public TokenFeeScheduleUpdateHandler(@NonNull final CustomFeesValidator customFeesValidator) {
+        this.customFeesValidator = customFeesValidator;
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java:57]: Done

File: hedera-node/hedera-token-service/src/main/java/com/hedera/node/app/service/token/ReadableTokenRelationStore.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ * <p>This class is not exported from the module. It is an internal implementation detail.
+ */
+public interface ReadableTokenRelationStore {
+    /**
+     * Returns the {@link TokenRelation} with the given number. If no such token relation exists, returns {@code Optional.empty()}
+     *
+     * @param tokenNum - the number of the token relation to be retrieved
+     * @param accountNum - the number of the account relation to be retrieved
+     */
+    Optional<TokenRelation> get(final long tokenNum, final long accountNum);
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service/src/main/java/com/hedera/node/app/service/token/ReadableTokenRelationStore.java:34]: Will do here as well.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java [hunk 8]
```diff
@@ -63,14 +81,67 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
...
+        // get the latest configuration
+        final var config = context.getConfiguration().getConfigData(TokenServiceConfig.class);
+        var op = txn.tokenFeeScheduleUpdateOrThrow();
+
+        // validate checks in handle
+        final var token = validateSemantics(op, tokenStore, config);
+
+        // create readable stores from the context
+        final var readableAccountStore = context.createReadableStore(ReadableAccountStore.class);
+        final var readableTokenRelsStore = context.createReadableStore(ReadableTokenRelationStore.class);
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java:106]: As I understand you can do either ways. But, any readable store we can create using this approach so it is inside handle and dispatcher will not provide them.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/ReadableTokenRelationStoreImpl.java (hunk 3 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/ReadableTokenRelationStoreImpl.java:52]: Doing that here

**Review by tinker-michaelj — DISMISSED** [review]
LGTM, nice work 🥂

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFeeScheduleUpdateHandler.java:107]: Ok, thanks for clarifying!

**Review by mhess-swl — DISMISSED** [review]

**Review by mhess-swl — APPROVED** [review]

**Pull request merged**